src/main/java/cz/tstrecha/timetracker/service/impl/UserServiceImpl.java

```diff
@@ -89,21 +85,57 @@ public UserDTO createUser(UserRegistrationRequestDTO registrationRequest, UserRo
         return userMapper.toDTO(savedUser);
     }
 
-    public RelationshipDTO createRelationship(RelationshipCreateUpdateRequestDTO request){
+    public RelationshipDTO createRelationship(RelationshipCreateUpdateRequestDTO request) {
+        var from = userRepository.findById(request.getFromId())
+                .orElseThrow(() -> new EntityNotFoundException("User entity not found by from id [" + request.getFromId() + "]"));
+        var to = userRepository.findById(request.getToId())
+                .orElseThrow(() -> new EntityNotFoundException("User entity not found by to id [" + request.getToId() + "]"));
+
+        if (userRelationshipRepository.existsByFromAndTo(from, to)) {
+            throw new UserInputException("Relationship already exists");
+        }
+        var relation = relationshipMapper.fromRequest(request, from, to);
+        relation = userRelationshipRepository.save(relation);
+        return relationshipMapper.toDTOFromReceiving(relation);
+    }
+
+    public RelationshipDTO createRelationship(RelationshipCreateUpdateRequestDTO request, LoggedUser loggedUser, UserContext userContext) {
+        if (!Objects.equals(userContext.getId(), loggedUser.getId()) || !Objects.equals(userContext.getId(), request.getFromId())) {
+            throw new UserInputException("You cannot create a relationship between other users");
+        }
+        return createRelationship(request);
+    }
+
+    public RelationshipDTO updateRelationShip(RelationshipCreateUpdateRequestDTO request, LoggedUser loggedUser, UserContext userContext) {
+        if (!userRelationshipRepository.existsById(request.getId())) {
+            throw new UserInputException("Relationship with this id doesn't exist");
+        }
+        if (!Objects.equals(userContext.getId(), loggedUser.getId())) {
+            throw new UserInputException("You cannot edit a relationship between other users");
+        }
+        if (!Objects.equals(userContext.getId(), userRelationshipRepository.findById(request.getId()).orElseThrow().getFrom().getId())) {
+            throw new UserInputException("You cannot edit a relationship between other users");
+        }
+        if (!Objects.equals(userContext.getId(), request.getFromId())) {
+            throw new UserInputException("You cannot edit a relationship between other users");
+        }
+
         var from = userRepository.findById(request.getFromId())
                 .orElseThrow(() -> new EntityNotFoundException("User entity not found by from id [" + request.getFromId() + "]"));
         var to = userRepository.findById(request.getToId())
                 .orElseThrow(() -> new EntityNotFoundException("User entity not found by to id [" + request.getToId() + "]"));
 
         var relation = relationshipMapper.fromRequest(request, from, to);
```

Comment:
The whole mapping thing here... it's not terrible, but it's not how its supposed to be:

you should find the entity, make a new mapping method

`public void updateRelationship(RequestDTO source, @MappingTarget RelationshipEntity target)` 

This should declare some `@Mapping(source =..., ignore = true)` ie for `from`, `to`, and `id`.

Reply:
Once you have done this, you dont really need to find from and to in repository (see @123-@126), because these entities are being held in the retrieved relationship entity.

Some checks will not be needed as well.